### ISSUE TITLE
Actions should probably be an enum now.

### ISSUE THREAD

**Issue by drvinceknight** [Axelrod-Python/Axelrod#816]
@marcharper you've mentioned this a couple of times, opening this issue to discuss :)

**Comment by marcharper** [comment]
It causes a lot of errors since `C` becomes `Actions.C` but I'm sure we can figure it out...

**Comment by marcharper** [comment]
Also note that `enum.auto()` is Python 3.6+.

**Comment by drvinceknight** [comment]
> It causes a lot of errors since C becomes Actions.C but I'm sure we can figure it out...

And perhaps from a readability point of view just having `C` is still preferable?

I am not actually familiar with `enum` at all. What problem would moving to them solve?

**Comment by marcharper** [comment]
> What problem would moving to them solve?
None that I have in mind at the moment but it is an Enum technically.

**Comment by eric-s-s** [comment]
here's my 10 minute demo of what it would look like.  for convenience, I built it in another project, but you get the idea.
```
from enum import Enum


class Action(Enum):
    C = 'C'
    D = 'D'

    def flip(self):
        flipped = {self.C: self.D, self.D: self.C}
        return flipped[self]

    def __repr__(self):
        return 'Action.{}'.format(self.name)
```

as far as i can see, the advantages are that 

- it's impossible to get an error when you flip.  
- It can't, in any way be mixed up with strings.  
- It now works in the type hints.  (i'll put the type hints for current action below.)  
- there is no need for Action and Actions to take care of type checking.  
- flip_action no longer needs a specific import.  It's just there for any C or D whenever you need it.

here's what it did in my interperter.
```
In[29]: from battle.scrap import Action
In[30]: C, D = Action.C, Action.D

In[32]: C
Out[32]:
Action.C

In[33]: C == C
Out[33]:
True

In[34]: C != D
Out[34]:
True

In[35]: C == "C"
Out[35]:
False

In[36]: C.flip()
Out[36]:
Action.D

In[37]: D.flip()
Out[37]:
Action.C

In[38]: def blah(num: int) -> Action:
   ...:     if num < 0:
   ...:         return C
   ...:     return D
   ...: 
In[39]: help(blah)
Help on function blah in module __main__:

blah(num:int) -> battle.scrap.Action

In[40]: blah(2)
Out[40]: 
Action.D

In[41]: x = 5 * [C]
In[42]: x
Out[42]: 
[Action.C, Action.C, Action.C, Action.C, Action.C]  <- not represented as strings
```

compare the type hint to the type hint i get in axelrod now
```
In[3]: import axelrod
In[4]: from axelrod.actions import flip_action
In[5]: help(flip_action)
Help on function flip_action in module axelrod.actions:

flip_action(action:str) -> str
```

**Comment by eric-s-s** [comment]
you could also make a scoring method.
C.score(C) returns (3, 3)
C.score(D) returns ....

**Comment by marcharper** [comment]
Did you try subbing this into axelrod? When I tried it broke **a lot** of tests.

**Comment by marcharper** [comment]
Could also add `random_choice(p)` to the class.

**Comment by eric-s-s** [comment]
not as yet.  I could put some actual work into it, but i'd like to finish with the response_test to versus_test first #884 .  when i run the test suites, i get snowed under by all the warnings generated by response_test.  (I didn't even know that's what was causing it until i started coding the first one in the issue.  as someone new to this package i was grateful for the change.  running those tests and not knowing to expect all the warnings confused the heck out of me.)

your from string comment in #925 could/should also be solved here.  Action.from_string('C') 

for all the tests it's breaking, it might be worth doing.  it IS an Enum that should own some important methods and where the code doesn't reflect that probably indicates things worth refactoring.

**Comment by marcharper** [comment]
I agree that it's worth doing, it's just going to be a lot of tests to fix. And yes we're trying to reduce the warnings -- you can suppress them all with the warnings package. The reason for the warnings is that some of the old tests used impossible histories as a shortcut to test strategy behavior, and while that works, it's not really the right way to do things.

**Comment by drvinceknight** [comment]
This all looks great! :)

> you could also make a scoring method.

Small point: I don't immediately see if this would be helpful as the scoring is currently the job of the `axelrod.Game` class which is used in the `interaction_utils` etc. We could pass a `game` instance to such a scoring method but that feels weird to me (could just be that it's different).

**Comment by drvinceknight** [comment]
> running those tests and not knowing to expect all the warnings confused the heck out of me.)

I've opened #930, with a one line addition to mute all the warnings. My thought process being that we're actively working to refactor the tests so they're done properly now so perhaps we can make things less "suprising". I'm not at all insistent on that though, @marcharper if you'd prefer to keep the warnings feel free to just close the PR.

**Comment by drvinceknight** [comment]
Have closed it, my one line broke other tests. Not worth the bother :)

**Comment by meatballs** [comment]
I don't like the scoring method as I think that's better done in the Game class, but I like the idea overall.

It might also be good to define the `__nonzero__` method so we can use boolean expressions on actions. (We'd have to decide which it True and which False).

**Comment by eric-s-s** [comment]
> I don't like the scoring method as I think that's better done in the Game class, but I like the idea overall.

👍 

ran the tests, finally.  a decent number of the failures just look like they can be managed by fixing areas where things assume Actions.C = 'C'.  this could be done in a series of small PR's that change a few things here or there, but don't break any original code.  

in that vein, it would be nice to have a temporary actions.py like so:

```
Action = str


class Actions(object):
    C = 'C'
    D = 'D'


# class Actions(Enum):
#     C = 'something'
#     D = 'or other'
# 
# Action = Actions
```

and a test_actions.py starting with:

```
class TestAction(unittest.TestCase):

    def test_action_values(self):
        if not isinstance(C, Actions):
            self.assertEqual(C, 'C')
            self.assertEqual(D, 'D')
            self.assertNotEqual(D, 'd')
            self.assertNotEqual(C, 'c')
            self.assertNotEqual(D, 0)
            self.assertNotEqual(C, 1)
            self.assertNotEqual(C, D)
```

that way, you could comment in/out the appropriate actions, find failures, fix a few, undo commenting, make sure nothing is broken and submit a PR.

at some point, you'd have to make a commitment one way or the other, but a not small amt could be changed like this.

**Comment by drvinceknight** [comment]
> ran the tests, finally. a decent number of the failures just look like they can be managed by fixing areas where things assume Actions.C = 'C'.

Naive question: could this be taken care of by a big find/replace on the test files in one PR?

**Comment by eric-s-s** [comment]
@drvinceknight 
it was a good question.  the short answer is, "AHHHHHHHH!"

the longer answer is, i don't think it did much. but it did break some stuff. not too much, but it barely helped either.

the real big help was changing test_player.TestPlayer   
first_play_test and second_play_test needed tweaking that make them a little ugly.  they call test_responses, which has a "responses" variable that is iterated over.  since a string is iterable, it never minded getting passed a single string.  but as an Enum, it needs to be encased in a list.

**Comment by eric-s-s** [comment]
now i didn't try it on the docs and i didn't run any doctests

**Comment by marcharper** [comment]
One big PR is preferable to multiple PRs that break some tests. Consider using the string_to_actions function that was recently added in `test_responses` (perhaps in conjunction with `isinstance(..., str)` to reduce the number of failing tests, and perhaps issue a warning when you do find a string of actions.

**Comment by eric-s-s** [comment]
i will happily try to tackle writing a PR that fixes test so that they're not broken in either implementation after finished helping  in #884 .   of course, if someone wants to go for it before then, please do.

**Comment by eric-s-s** [comment]
hello all,

I have free time again and am interested in tackling this issue.

below is code that incorporates roughly what was being discussed.  i notice there was no discussion to call it Action or Actions.  Let me know what you think - what should be set and how, and if it's still a good idea to go ahead with this.  

```
class Action(Enum):
    C = 1
    D = 0
    def __bool__(self):
        return bool(self.value)

    def __repr__(self):
        return 'Action.{}'.format(self.name)

    def flip(self):
        flipped = {self.C: self.D, self.D: self.C}
        return flipped[self]

    @classmethod
    def random_choice(cls, p: float = 0.5) -> 'Action':      
        if p == 0:
            return cls.D
        if p == 1:
            return cls.C
        r = random.random()
        if r < p:
            return cls.C
        return cls.D


def flip_action(action: Action) -> Action:
    if not isinstance(action, Action):
        raise ValueError("Encountered an invalid action.")
    return action.flip()


def str_to_actions(actions: str) -> tuple:
    action_dict = {'C': Action.C,
                   'D': Action.D}
    try:
        return tuple(action_dict[action] for action in actions)
    except KeyError:
        raise ValueError('The characters of "actions" str may only be "C" or "D"')
```

**Comment by marcharper** [comment]
@eric-s-s Welcome back! Thanks for tackling this. IIRC we've finished the test conversions so there should not be any warnings in the way.

For efficiency I might implement str_to_actions as:

```
def str_to_actions(actions: str) -> tuple:
    for s in str:
        if s == 'C':
             yield Action.C
        if s == 'D':
             yield Action.D
        else:
             raise UnknownAction #or similar
```

I also recommend implementing `flip` as something like

```
def flip(self, action):
    if self == Action.C:
        return Action.D
    if self == Action.D:
        return Action.C
    else:
        raise UnknownAction(...)
```

and 

```
    def __repr__(self):
        return '{}'.format(self.name)
```

Whenever you are ready please open a PR -- hopefully not too many tests are affected.

**Comment by eric-s-s** [comment]
@marcharper i like those. i think the first time we talked about `str_to_actions` that was the code.  The only issue with a generator vs tuple is that the tuple fails immediately and the generator only fails once the action is reached.  With a generator,  `Cycler('CCoops')`  doesn't fail until round 3 while the tuple(style?) fails at instantiation. 

str_to_actions is currently only used in two places though.  

I'll start in on this at a decent pace, and that's a very easy detail to tweak.

**Comment by marcharper** [comment]
Ok, tuple is fine instead of a generator. I think hard coding the logic is probably faster than building a dictionary every time; regardless if we find a more efficient implementation we can also refactor, the bigger issue is making sure all the tests pass for the enum.

**Comment by eric-s-s** [comment]
from [test_player](https://github.com/Axelrod-Python/Axelrod/blob/master/axelrod/tests/strategies/test_player.py#L556)

```
def test_four_vector(test_class, expected_dictionary):
    """
    Checks that two dictionaries match -- the four-vector defining
    a memory-one strategy and the given expected dictionary.
    """
    player1 = test_class.player()
    for key in sorted(expected_dictionary.keys()):
        test_class.assertAlmostEqual(
            player1._four_vector[key], expected_dictionary[key])
```

does anyone have any idea why the dictionary keys are sorted?  i removed the `sorted` to no ill effect on tests.  or i can leave the `sorted` and write `__lt__`, `__gt___` etc.. for Actions.

**Comment by eric-s-s** [comment]
okay, now that the Enum is in, these two seem like they would be nice payoffs:

1 - remove `flip_action` . since actions are type hinted, any instance of that can be replaced by `action.flip()`. It would also remove any need for error checking
2 - remove the alias `Action = Actions` and either have `Action.C` and `Action.D` are of type `Action` from the module `action.py` OR `Actions.C` and `Actions.D` are of the type `Actions` from the module `actions.py`

I fiddled with it in a branch and they are relatively easy changes to make with an IDE and sed.  If people like these, i'll proceed (as two different PRs if they both meet approval).  If you like pt. 2, i'll wait on a decision on what to call it.

**Comment by marcharper** [comment]
@eric-s-s sounds good to me

**Comment by eric-s-s** [comment]
Next I would love to remove the alias.  Should the class name be `Action` or `Actions`?

**Comment by drvinceknight** [comment]
> Should the class name be Action or Actions?

It should be `Action` I reckon.

**Comment by eric-s-s** [comment]
if the class name is `Action`, should i leave the module name as `actions.py` or also change that to `action.py` ?

**Comment by eric-s-s** [comment]
Is there more to do with this issue, or is it now closable?

**Comment by drvinceknight** [comment]
Nope I think this is closable: thanks @eric-s-s!! :+1:

**Comment by eric-s-s** [comment]
you're welcome.  i enjoyed doing it.
